[Flaky-test] DistributedClusterTest.testMutiBrokerUnloadReload

`DistributedClusterTest` is flaky. The `testMutiBrokerUnloadReload` test method fails sporadically.

[example failure](https://github.com/streamnative/kop/runs/3396339243?check_suite_focus=true#step:5:167)

```
Error:  testMutiBrokerUnloadReload(io.streamnative.pulsar.handlers.kop.DistributedClusterTest)  Time elapsed: 11.62 s  <<< FAILURE!
java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.TimeoutException: Expiring 1 record(s) for kopMutiBrokerUnloadReload10-0: 10001 ms has passed since batch creation plus linger time
	at org.apache.kafka.clients.producer.internals.FutureRecordMetadata.valueOrError(FutureRecordMetadata.java:94)
	at org.apache.kafka.clients.producer.internals.FutureRecordMetadata.get(FutureRecordMetadata.java:64)
	at org.apache.kafka.clients.producer.internals.FutureRecordMetadata.get(FutureRecordMetadata.java:29)
	at io.streamnative.pulsar.handlers.kop.DistributedClusterTest.kafkaPublishMessage(DistributedClusterTest.java:197)
	at io.streamnative.pulsar.handlers.kop.DistributedClusterTest.testMutiBrokerUnloadReload(DistributedClusterTest.java:424)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:54)
	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:44)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.kafka.common.errors.TimeoutException: Expiring 1 record(s) for kopMutiBrokerUnloadReload10-0: 10001 ms has passed since batch creation plus linger time

```

--- a/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java
+++ b/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java
@@ -2258,18 +2258,17 @@ public class KafkaRequestHandler extends KafkaCommandDecoder {
         CompletableFuture<PartitionMetadata> returnFuture = new CompletableFuture<>();
 
         topicManager.getTopicBroker(topic.toString())
-                .thenCompose(address -> getProtocolDataToAdvertise(address, topic))
-                .whenComplete((stringOptional, throwable) -> {
-                    if (stringOptional == null || !stringOptional.isPresent() || throwable != null) {
-                        log.error("Not get advertise data for Kafka topic:{}. throwable: [{}]",
-                            topic, throwable.getMessage());
+                .thenApply(address -> getProtocolDataToAdvertise(address, topic))
+                .thenAccept(kopAddressFuture -> kopAddressFuture.thenAccept(listenersOptional -> {
+                    if (!listenersOptional.isPresent()) {
+                        log.error("Not get advertise data for Kafka topic:{}.", topic);
                         KafkaTopicManager.removeTopicManagerCache(topic.toString());
                         returnFuture.complete(null);
                         return;
                     }
 
                     // It's the `kafkaAdvertisedListeners` config that's written to ZK
-                    final String listeners = stringOptional.get();
+                    final String listeners = listenersOptional.get();
                     final EndPoint endPoint =
                             (tlsEnabled ? EndPoint.getSslEndPoint(listeners)
                                     : EndPoint.getPlainTextEndPoint(listeners));
@@ -2277,8 +2276,8 @@ public class KafkaRequestHandler extends KafkaCommandDecoder {
 
                     if (log.isDebugEnabled()) {
                         log.debug("Found broker localListeners: {} for topicName: {}, "
-                                + "localListeners: {}, found Listeners: {}",
-                            listeners, topic, advertisedListeners, listeners);
+                                        + "localListeners: {}, found Listeners: {}",
+                                listeners, topic, advertisedListeners, listeners);
                     }
 
                     // here we found topic broker: broker2, but this is in broker1,
@@ -2287,7 +2286,13 @@ public class KafkaRequestHandler extends KafkaCommandDecoder {
                         KafkaTopicManager.removeTopicManagerCache(topic.toString());
                     }
                     returnFuture.complete(newPartitionMetadata(topic, node));
-            });
+                })).exceptionally(throwable -> {
+                    log.error("Not get advertise data for Kafka topic:{}. throwable: [{}]",
+                            topic, throwable.getMessage());
+                    KafkaTopicManager.removeTopicManagerCache(topic.toString());
+                    returnFuture.complete(null);
+                    return null;
+                });
         return returnFuture;
     }
 

--- a/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java
+++ b/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java
@@ -2258,34 +2258,35 @@ public class KafkaRequestHandler extends KafkaCommandDecoder {
         CompletableFuture<PartitionMetadata> returnFuture = new CompletableFuture<>();
 
         topicManager.getTopicBroker(topic.toString())
-            .thenCompose(address -> getProtocolDataToAdvertise(address, topic))
-            .whenComplete((stringOptional, throwable) -> {
-                if (!stringOptional.isPresent() || throwable != null) {
-                    log.error("Not get advertise data for Kafka topic:{}. throwable: [{}]",
-                        topic, throwable.getMessage());
-                    KafkaTopicManager.removeTopicManagerCache(topic.toString());
-                    returnFuture.complete(null);
-                    return;
-                }
+                .thenCompose(address -> getProtocolDataToAdvertise(address, topic))
+                .whenComplete((stringOptional, throwable) -> {
+                    if (stringOptional == null || !stringOptional.isPresent() || throwable != null) {
+                        log.error("Not get advertise data for Kafka topic:{}. throwable: [{}]",
+                            topic, throwable.getMessage());
+                        KafkaTopicManager.removeTopicManagerCache(topic.toString());
+                        returnFuture.complete(null);
+                        return;
+                    }
 
-                // It's the `kafkaAdvertisedListeners` config that's written to ZK
-                final String listeners = stringOptional.get();
-                final EndPoint endPoint =
-                        (tlsEnabled ? EndPoint.getSslEndPoint(listeners) : EndPoint.getPlainTextEndPoint(listeners));
-                final Node node = newNode(endPoint.getInetAddress());
+                    // It's the `kafkaAdvertisedListeners` config that's written to ZK
+                    final String listeners = stringOptional.get();
+                    final EndPoint endPoint =
+                            (tlsEnabled ? EndPoint.getSslEndPoint(listeners)
+                                    : EndPoint.getPlainTextEndPoint(listeners));
+                    final Node node = newNode(endPoint.getInetAddress());
 
-                if (log.isDebugEnabled()) {
-                    log.debug("Found broker localListeners: {} for topicName: {}, "
-                            + "localListeners: {}, found Listeners: {}",
-                        listeners, topic, advertisedListeners, listeners);
-                }
+                    if (log.isDebugEnabled()) {
+                        log.debug("Found broker localListeners: {} for topicName: {}, "
+                                + "localListeners: {}, found Listeners: {}",
+                            listeners, topic, advertisedListeners, listeners);
+                    }
 
-                // here we found topic broker: broker2, but this is in broker1,
-                // how to clean the lookup cache?
-                if (!advertisedListeners.contains(endPoint.getOriginalListener())) {
-                    KafkaTopicManager.removeTopicManagerCache(topic.toString());
-                }
-                returnFuture.complete(newPartitionMetadata(topic, node));
+                    // here we found topic broker: broker2, but this is in broker1,
+                    // how to clean the lookup cache?
+                    if (!advertisedListeners.contains(endPoint.getOriginalListener())) {
+                        KafkaTopicManager.removeTopicManagerCache(topic.toString());
+                    }
+                    returnFuture.complete(newPartitionMetadata(topic, node));
             });
         return returnFuture;
     }

--- a/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopBrokerLookupManager.java
+++ b/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopBrokerLookupManager.java
@@ -68,7 +68,7 @@ public class KopBrokerLookupManager {
         getTopicBroker(topic)
                 .thenCompose(pair -> getProtocolDataToAdvertise(pair, TopicName.get(topic)))
                 .whenComplete((listeners, throwable) -> {
-                    if (!listeners.isPresent() || throwable != null) {
+                    if (listeners == null || !listeners.isPresent() || throwable != null) {
                         log.error("Not get advertise data for Kafka topic:{}. throwable", topic, throwable);
                         returnFuture.complete(null);
                         return;

--- a/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopBrokerLookupManager.java
+++ b/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopBrokerLookupManager.java
@@ -66,10 +66,11 @@ public class KopBrokerLookupManager {
 
         CompletableFuture<InetSocketAddress> returnFuture = new CompletableFuture<>();
         getTopicBroker(topic)
-                .thenCompose(pair -> getProtocolDataToAdvertise(pair, TopicName.get(topic)))
-                .whenComplete((listeners, throwable) -> {
-                    if (listeners == null || !listeners.isPresent() || throwable != null) {
-                        log.error("Not get advertise data for Kafka topic:{}. throwable", topic, throwable);
+                .thenApply(address -> getProtocolDataToAdvertise(address, TopicName.get(topic)))
+                .thenAccept(kopAddressFuture -> kopAddressFuture.thenAccept(listeners -> {
+                    if (!listeners.isPresent()) {
+                        log.error("Not get advertise data for Kafka topic:{}", topic);
+                        removeTopicManagerCache(topic);
                         returnFuture.complete(null);
                         return;
                     }
@@ -93,6 +94,12 @@ public class KopBrokerLookupManager {
                     }
 
                     checkTopicOwner(returnFuture, topic, endPoint);
+                })).exceptionally(throwable -> {
+                    log.error("Not get advertise data for Kafka topic:{}. throwable: [{}]",
+                            topic, throwable.getMessage());
+                    removeTopicManagerCache(topic);
+                    returnFuture.complete(null);
+                    return null;
                 });
         return returnFuture;
     }

--- a/tests/src/test/java/io/streamnative/pulsar/handlers/kop/DistributedClusterTest.java
+++ b/tests/src/test/java/io/streamnative/pulsar/handlers/kop/DistributedClusterTest.java
@@ -381,9 +381,7 @@ public class DistributedClusterTest extends KopProtocolHandlerTestBase {
     }
 
     // Unit test for unload / reload user topic bundle, verify it works well.
-    // NOTE: Currently the testMultiBrokerUnloadReload is flaky. If it ran after other tests, it would be easy to fail.
-    //   So we just change the priority to make it run first to avoid CI failing at this test for this moment.
-    @Test(timeOut = 30000, priority = -1)
+    @Test(timeOut = 30000)
     public void testMultiBrokerUnloadReload() throws Exception {
         int partitionNumber = 10;
         String kafkaTopicName = "kopMultiBrokerUnloadReload" + partitionNumber;

